synapse/handlers/room.py

```diff
@@ -517,21 +512,30 @@ def _move_aliases_to_new_room(
                 # checking module decides it shouldn't, or similar.
                 logger.error("Error adding alias %s to new room: %s", alias, e)
 
+        # The canonical alias and alternative aliases can only be propagated to
+        # the new room if they were removed from the old room.
+        content = {}
+        if canonical_alias and (canonical_alias in removed_aliases):
+            content["alias"] = canonical_alias
+        canonical_alt_aliases = [
+            alias for alias in canonical_alt_aliases if alias in removed_aliases
+        ]
+        if canonical_alt_aliases:
+            content["alt_aliases"] = canonical_alt_aliases
+
```

Comment:
I think this is incorrect wrt our conversation earlier

Reply:
I've modified this code to pass through the previous canonical event's content unchanged.